keps/0032-community-repository-management.md

```diff
@@ -0,0 +1,96 @@
+- Some operator tests cannot be covered by the test tooling provided in `kudobuilder/operators` making it necessary to host the tests in a separate Git repository.
+- Providing multiple versions of an operator is only possible by using separate folders. Teams developing operators might prefer different workflows (e.g. Git tags)
+
+These requirements make it necessary for operator packages to be developed in a separate Git repository. As a result, the operator packages in `kudobuilder/operators` are copies of specific versions of the respective upstream Git repository. This approach has challenges as well, because bugs discovered in an operator need to be resolved in the upstream Git repository, not in `kudobuilder/operators`. I.e., there needs to be metadata to link to the upstream Git repository. By having some form of metadata describing upstream operator package sources, the same metadata can be used to describe other properties of a package, e.g. the maturity level.
+
+### Goals
+
+- Provide a simple workflow for upstream operator packages to get added to the community repository
+- Remove the need to "host" copies of upstream operator packages in a Git repository
+- Provide a mechanism to define upstream operator package sources
+- Provide a mechanism to define upstream operator package maturity
+
+### Non-Goals
+
+- Change the practice of creating a PR against a Git repository to add operator packages to the community repository
+- Change operator package management as defined in [KEP-10](0010-package-manager.md) and [KEP-15](0015-repository-management.md)
+
+## Proposal
+
+Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. For example, consider an operator package developed at `github.com/example/example-operator` that has a tagged version `1.0.0` and the operator package in the `operator` folder. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like
+
+```yaml
+name: Example Operator
+versions:
+- version: "1.0.0"
+  git:
+    repository: github.com/example/example-operator.git
+    tag: "1.0.0"
+    dir: operator
+```
+
+Once this PR is merged, CI tooling detects the new YAML file, clones the referenced upstream Git repository, checks out the tag, and adds the operator package in the specified folder to the existing index. Of course, CI tests that don't update the community repository can run before the PR is merged. This workflow is similar to [krew-index](https://github.com/kubernetes-sigs/krew-index)
+
+We can add more metadata to the YAML reference file. E.g., support for different upstream sources like Mercurial.
```

Comment:
Should we want to generate a web presence for the operators (e.g. like Helm Hub) we'd need more info e.g. maintainer info (URL and whatnot) and operator description (README, installation instructions, list of params, that sort of things). Should this be part of the upstream repo?

Reply:
To be discussed. Ideally, all of this information should be provided as part of the operator package.